### ISSUE TITLE
[bug]: scammed using a hold invoice

### ISSUE THREAD

**Issue by anycolo** [lightningnetwork/lnd#7621]
1. I tried paying a ln invoice using my lnd instance. I submitted the payment a few days ago, but the status i received wasn't "complete". 
2. I waited a few hours, i tried paying again, i got 'no route' or 'payment in transit' (not sure)
3. I waited a few more hours, i tried paying again, it **didn't** say 'invoice is already paid'.
4. I waited a few moure hours, i tried paying the invoice again, it said that it was EXPIRED, so i assumed that the payment failed.

I refunded the customer using another channel and .. behold, the invoice managed to complete itself after some time, so i paid him twice.

I have a lot of questions:
- was it possible, that between 3 and 4 the payment completed and when trying to pay it again, lnd woulnd't respond with 'invoice is already paid' but with 'invoice expired'?
- i am 99% sure that this was a hold invoice. is it possible to force lnd to pay an invoice in X amount of time or cancel it, with me being able to retry it later and get a clear status of what is going on?
- can an invoice be settled AFTER the expiration date, using these stuck HTLC's? If so, why? It seems like a very bad idea.
- how can lnd cancel a transaction that is in flight? how can i cancel a hodl invoice?
- how can i set the max amount of time a HTLC can get stuck by the other side?

**Comment by alexbosworth** [comment]
payment in transit means that the payment is being sent, but it shouldn't be possible that you can pay the same payment request twice in lnd, probably need more logs or more details to know exactly what happened

**Comment by anycolo** [comment]
I didn't pay the same invoice twice, i refunded via classic bitcoin transaction.

**Comment by alexbosworth** [comment]
ok, just to note, yes an invoice that is past the "expiry" date marked on the payment request can still be settled after this date. The date is advisory and not binding

**Comment by anycolo** [comment]
Is this behavior a bug in lightning, in general?

**Comment by anycolo** [comment]
I have a payment request "lnbcxxxx..." from yesterday. My lnd wallet attempted to pay it a few times. What command can i run to see:
- if it was attempted
- what was the result of that attempt(s)
- if the status isn't clear, how can i stop any future attempts
- if there are any stuck htlcs for that payment (part of an hodl invoice)

**Comment by alexbosworth** [comment]
you can use trackpayment

**Comment by anycolo** [comment]
Trackpayment requires a hash, so i have decoded the payment request with decodepayreq, then i ran trackpayment succesfully:

```
+------------+--------------+--------------+--------------+-----+----------+----------+-------+
| HTLC_STATE | ATTEMPT_TIME | RESOLVE_TIME | RECEIVER_AMT | FEE | TIMELOCK | CHAN_OUT | ROUTE |
+------------+--------------+--------------+--------------+-----+----------+----------+-------+
+------------+--------------+--------------+--------------+-----+----------+----------+-------+
Amount + fee:   0 + 0 sat
Payment hash:   xxx
Payment status: FAILED, reason: FAILURE_REASON_NO_ROUTE

```

Does this mean 100% sure that the payment will not be completed at a later time?

**Comment by alexbosworth** [comment]
yes

**Comment by anycolo** [comment]
Okay, assume this case again:
- attempt to pay an invoice
- payment is held hostage by HODL invoice
- invoice expires
- attempt to pay the invoice again to see it's status

What should lnd respond when i try to do this:
- A: payment is in transit
- B: invoice has expired
- C: something else?

**Comment by alexbosworth** [comment]
if the htlc is still locked then it should respond that the payment is in transit yes, it shouldn't let you double pay it

**Comment by anycolo** [comment]
Okay, but any way to cancel a 'payment in transit' invoice? Or do i have to wait for the HTLC to resolve?

**Comment by alexbosworth** [comment]
Right there is a timeout associated with HTLCs that are in flight, wait for that timeout to expire

**Comment by anycolo** [comment]
I notice that sometimes, i can submit a payment attempt to my LND node, and it will timeout with some error after a few minutes of trying. When i try to make the same payment in a few minutes, i'll get an 'invoice is already paid'. How can i instruct lnd to cancel a submitted payment attempt or how can i instruct LND to try paying it for only a finite amount of time and then abort forever. It seems that once you submit an invoice to pay to lnd, it goes into a grey zone where you can never be sure about it's status.

**Comment by alexbosworth** [comment]
you couldn't cancel a payment with already committed HTLCs, but you can specify a pathfinding timeout to payment

You also cannot cancel pathfinding within that timeout frame though

The status can be found via trackpayment api

**Comment by HenryEnrique** [comment]
> Okay, but any way to cancel a 'payment in transit' invoice? Or do i have to wait for the HTLC to resolve?

Same thing happened to me but thanks to Blackhathacker111 who helped me get recovered my lost funds,message him on telegram his telegram username is https://t.me/Blackhathacker110

**Comment by ZZiigguurraatt** [comment]
Is this really a bug? I'm thinking we should close it as it seems to be just a misunderstanding on how invoices and HTLC work.

**Comment by saubyk** [comment]
Agree. closing this issue, can be reopened if more clarification is needed.

**Comment by anycolo** [comment]
IMHO, attempting to pay an expired invoice WHICH was settled should warn the user that it was paid:
- expired but paid
- expired but not paid
- expired but in transit (?)
Having invoices with an expiration date makes users think that there is 0% chance of that invoice settling after the expiration time, which is not actually true.

**Comment by ZZiigguurraatt** [comment]
Are you able to reliably reproduce your issue? I have not heard of anyone else experiencing the same problem.

**Comment by ZZiigguurraatt** [comment]
If you are wanting to cancel a HOLD invoice after you've already sent it, that is not possible in the protocol.

I think you are wanting a HOLD invoice to expire when the invoice expires, but you can't do that, you need to wait for the HTLC to expire, which is usually much longer than the invoice expiry. The invoice expiry is how long you can wait before sending the payment, but it isn't how long until the payment must be settled by. The `min_final_cltv_expiry` in the invoice sets that.

If you are saying you can pay the same HOLD invoice twice before the invoice expires, please clarify that and make sure it is repeatable.

**Comment by ZZiigguurraatt** [comment]
If you are saying that if you try to pay an already paid invoice, but aren't getting a clear error message when it fails and that is what you want, also please clarify that.

**Comment by anycolo** [comment]
- i now know it's not possible to cancel a hold invoice.
- yes, correct. but i'm talking about the way you interact with lnd regarding that invoice.
- no, i am not saying you can pay a hold invoice twice.

here is what i am saying:
- i trade goods for bitcoin
- the buyer makes a hold invoice which expires at 12:00 today
- i pay the invoice at 09:00 today
- the invoice is nearing 12:00 but still isn't settled
- at 11:59 i try to pay the invoice again, it says it's in transit
- at 12:01 i try to pay the invoice again, it says it's expired
- i now assume the payment towards him has failed, so i refund the goods
- after i refund the goods back to him, as i wasn't able to pay, the buyer will unlock the hodl invoice and "rob" me.

- at 12:01, during the payment attempt, the error message should have been something like "the invoice expired, but the payment is in transition and may complete until .. x days in the future".

**Comment by ZZiigguurraatt** [comment]
> * the buyer makes a hold invoice which expires at 12:00 today

only the seller can create a HOLD invoice. the buyer can't create the invoice.

**Comment by ZZiigguurraatt** [comment]
> * the invoice is nearing 12:00 but still isn't settled

if you've paid the invoice but the seller hasn't settled yet, it is not your problem. you've done everything you can do. if they don't settle it, they are choosing not to accept payment that you've sent.

**Comment by ZZiigguurraatt** [comment]
> * at 11:59 i try to pay the invoice again, it says it's in transit

I would think this is expected because you can't pay an invoice twice.

**Comment by ZZiigguurraatt** [comment]
> * at 12:01 i try to pay the invoice again, it says it's expired

this is by design, because you can't pay an invoice after it has expired, so LND probably doesn't even try. maybe we could make LND be smart enough to check if there is an unsettled payment before returning this error, even if the invoice is already expired?

**Comment by ZZiigguurraatt** [comment]
> * i now assume the payment towards him has failed, so i refund the goods

why would you refund goods if you are the one trying to send payment? this makes no sense.

**Comment by ZZiigguurraatt** [comment]
> * after i refund the goods back to him, as i wasn't able to pay, the buyer will unlock the hodl invoice and "rob" me.

I don't think you should ever return goods if you have an unsettled payment, even if the invoice has expired. at the same time, the merchant should never give the goods to you if they have not settled the payment. I think a big use case for HOLD invoices is for the merchant to be able to cancel the payment if they actually don't have any inventory left at the time you try to make payment. There could be 100 people with the goods in their shopping cart, but they only have inventory for 1 customer, but they have no idea who is going to actually pay until some payment is received. they can receive payment, and then the first customer that actually pays, then they have the payment made and they can cancel all the other people's shopping carts, then once they have done that and properly reserved the goods for the 1 customer, then accept the payment and deliver the goods. however, they have no motivation to cancel everyone else's shopping carts until they get some payment from someone. at the same time, if they receive a payment from two customers within a few seconds, before they have time to cancel everyone's shopping carts, they can cancel the second payment if they have no inventory left.

**Comment by anycolo** [comment]
I didn't know the invoice settled, this is why i refunded. I assumed it would never complete since it was expired.

**Comment by ZZiigguurraatt** [comment]
> I didn't know the invoice settled, this is why i refunded. I assumed it would never complete since it was expired.

I'd re-read https://github.com/lightningnetwork/lnd/issues/7621#issuecomment-2997181845 , you can also test and see that a hold invoice can be settled after the expiration time but before the lock time expiry.

The invoice expiry has no on chain enforcement as far as I know. It's mainly there to control if you will accept or reject a payment that is attempted, but then you have more time to settle it after the invoice expiry which is enforced on chain. That is assuming that the `min_final_cltv_expiry` is longer than the invoice expiry (which it always should be). One reason we have both types of expiry is we need to make sure we have enough time to do any on chain enforcement if that is required.
